extension/observer/ecsobserver/matcher.go

```diff
@@ -57,6 +57,24 @@ type MatchedContainer struct {
 	Targets        []MatchedTarget
 }
 
+// MergeTargets adds new targets to the set, the 'key' is port + metrics path.
+// The 'key' does not have ip because all the targets from on container has same ip.
+// If there are duplicated 'key' we honor existing target and does not override.
+// The duplication could happen if there are several rules matching same target.
+func (mc *MatchedContainer) MergeTargets(newTargets []MatchedTarget) {
+	for _, newt := range newTargets {
+		for _, old := range mc.Targets {
+			// If port and metrics_path are same, then we treat them as same target and keep the existing one
+			if old.Port == newt.Port && old.MetricsPath == newt.MetricsPath {
+				continue
+			}
+			mc.Targets = append(mc.Targets, newt)
+		}
+	}
```

Comment:
good catch,  it's not. I think we can change the label to `NextTarget` I don't recall if there are style guide for all caps goto labels. 

I will update the unit test as well.  

Those logic actually rarely got trigged in our [e2e test](https://github.com/aws-observability/aws-otel-test-framework) because most of sample workload config does not match same task using different matching rules and assign different port path combination. But since we don't allow multiple metrics path in one config, this logic is one way to workaround it. (btw: It's different from k8s discovery where all the ports are discovered and filtered later in relabel)

```
- ports:
    - 123
    - 456
  path: /metrics
- ports:
   - 789
   path: /another-metrics-path
```

The reason we don't allow config both is it's a bit confusing and try to be close to existing cloudwatch agent's ecs prometheus discovery config.

```
If both ports and paths are array

- ports:
   - 123
   - 456
   paths:
  - /metrics
  - /another-metrics

Will result in the following, which is not a common case for workloads in cloudwatch's container insight. 
But if it is common for other workloads, we may switch to this config pattern.

:123/metrics
:123/another-metrics
:456/metrics
:456/another-metrics
```

Reply:
The discovery ticker adds or merges the newly matched targets with the old/cached targets. Do we need to remove the old target from Task structure it is not matched anymore?

Reply:
We don't cache task, we only (LRU) cache task definition, ec2 instance etc. We don't save previous matched result in memory. Every tick we simply scan all the tasks, do the match and replace the content of entire file sd. Prometheus itself (discovery manger) will handle the diff logic and the discovery plugins can just do full dump. And there is no way for plugin like file sd to communicate with prometheus.

Reply:
Fixed in latest commit, also added some unit test.